app/models/campaign_page.rb

```diff
@@ -3,20 +3,28 @@ class CampaignPage < ActiveRecord::Base
 
   belongs_to :language
   belongs_to :campaign # Note that some campaign pages do not necessarily belong to campaigns
-   
-  has_many :campaign_pages_widgets, inverse_of: :campaign_page
-  has_many :campaign_pages_tags
-  has_many :tags, through: :campaign_pages_tags, source: :tag
+
+  has_many :campaign_pages_tags, dependent: :destroy
+  has_many :tags, through: :campaign_pages_tags
+  has_many :widgets, dependent: :destroy, as: :page
 
   validates :title, :slug, presence: true, uniqueness: true
+  validates :language, presence: true
 
   # validating presence of a boolean fields
   validates_inclusion_of :active, in: [true, false]
   validates_inclusion_of :featured, in: [true, false]
 
   # calls validations on the widgets associated to the campaign page:
-  validates_associated :campaign_pages_widgets
+  validates_associated :widgets
 
   # allows updating associated campaign page widgets
-  accepts_nested_attributes_for :campaign_pages_widgets
+  accepts_nested_attributes_for :widgets
+
+  before_validation :create_slug
+
+  # have we thought about using friendly id? probably better
```

Comment:
Probably makes sense since we'll want to do a lot of searching by slug. Opened #53.